ui/raidboss/data/06-ew/raid/p12s.ts

```diff
@@ -61,6 +61,24 @@ const superchainNpcBaseIdMap: Record<SuperchainMechanic, string> = {
   partners: '16180',
 } as const;
 
+const engravementIdMap = {
+  lightTilt: 'DF8',
+  darkTilt: 'DF9',
+  lightTower: 'DFB',
+  darkTower: 'DFC',
+  lightBeam: 'DFD',
+  darkBeam: 'DFE',
+  crossMarked: 'DFF',
+  xMarked: 'E00',
+};
+const engravementLabelMap = Object.fromEntries(
+  Object.entries(engravementIdMap).map(([k, v]) => [v, k]),
+);
```

Comment:
> Re: typing, you can do this dance. Inverting objects is always a little fraught in TypeScript because of the way Object.keys and Object.entries strips some of them types.

Oh, that is a neat trick.  What do you think about simplifying this by reversing the declarations and using a type annotation on fromEntries - I think that would let us type both objects?  E.g.:

```
type EngravementLabel = 'lightTilt' | 'darkTilt' | 'lightTower' | 'darkTower' | 'darkTower' | 'lightBeam' | 'darkBeam' | 'crossMarked' | 'xMarked';
const engravementLabelMap: { [effectId: string]: EngravementLabel } = {
  DF8: 'lightTilt',
  DF9: 'darkTilt',
  DFB: 'lightTower',
  DFC: 'darkTower',
  DFD: 'lightBeam',
  DFE: 'darkBeam',
  DFF: 'crossMarked',
  E00: 'xMarked',
} as const;
type EngravementIdMap = Record<EngravementLabel, string>;
const engravementIdMap: EngravementIdMap = Object.fromEntries(Object.entries(engravementLabelMap).map(([k, v]) => [v, k])) as EngravementIdMap;
```

Reply:
I think that's fine too.  I looked at this a bit and couldn't come up with any way to avoid the `as` type casting in either case.  I thought maybe `as` on the keys looked safer (because it was simpler and felt obvious), but I think this is also fine.

You should be able to still say: `type EngravementLabel = typeof engravementLabelMap[keyof typeof engravementLabelMap];` here without duplicating the labels in the first line as well.

Reply:
> You should be able to still say: `type EngravementLabel = typeof engravementLabelMap[keyof typeof engravementLabelMap];` here without duplicating the labels in the first line as well.

I tried that initially, and I think it leads to some sort of weird recursion error, which is why I duplicated the labels.  This is probably my still-rough knowledge of typing, but regardless of whether the `type EngravementLabel` definition comes before or after `const engravementLabelMap`, the const declaration throws an error because EngravementLabel is not validly defined at that point (i.e. 'Type 'string' is not assignable to type 'never'').

Reply:
```typescript
const engravementLabelMap = {
  DF8: 'lightTilt',
  DF9: 'darkTilt',
  DFB: 'lightTower',
  DFC: 'darkTower',
  DFD: 'lightBeam',
  DFE: 'darkBeam',
  DFF: 'crossMarked',
  E00: 'xMarked',
} as const;
type EngravementLabel = typeof engravementLabelMap[keyof typeof engravementLabelMap];
type EngravementIdMap = Record<EngravementLabel, string>;
const engravementIdMap: EngravementIdMap = Object.fromEntries(Object.entries(engravementLabelMap).map(([k, v]) => [v, k])) as EngravementIdMap;
```

With the `as const`, you don't have to explicitly type `engravementLabelMap`

Reply:
> With the `as const`, you don't have to explicitly type `engravementLabelMap`

I think that ends up breaking a bunch of the lookups against `engravementLabelMap[]` though.  Like `const engraveLabel = engravementLabelMap[matches.effectId];` is now an unsafe assignment of an 'any' value.

Reply:
The original thing I posted didn't have this because there's no lookups the other way around.  You can fix that by re-typing it after:

```typescript
const engravementLabelMapAsConst = {
  DF8: 'lightTilt',
  DF9: 'darkTilt',
  DFB: 'lightTower',
  DFC: 'darkTower',
  DFD: 'lightBeam',
  DFE: 'darkBeam',
  DFF: 'crossMarked',
  E00: 'xMarked',
} as const;
type EngravementLabel = typeof engravementLabelMapAsConst[keyof typeof engravementLabelMapAsConst];
const engravementLabelMap: { [effectId: string]: EngravementLabel } = engravementLabelMapAsConst;
type EngravementIdMap = Record<EngravementLabel, string>;
const engravementIdMap: EngravementIdMap = Object.fromEntries(Object.entries(engravementLabelMap).map(([k, v]) => [v, k])) as EngravementIdMap;
```

Reply:
Thanks!